rust/agama-server/src/web/http.rs

```diff
@@ -69,6 +72,33 @@ pub async fn login(
     Ok((headers, content))
 }
 
+#[derive(Clone, Deserialize, utoipa::ToSchema)]
+pub struct LoginFromQueryParams {
+    /// Token to use for authentication.
+    token: String,
+}
+
+#[utoipa::path(get, path = "/login", responses(
+    (status = 301, description = "Injects the authentication cookie if correct and redirects to the web UI")
+))]
+pub async fn login_from_query(
+    State(state): State<ServiceState>,
+    Query(params): Query<LoginFromQueryParams>,
+) -> impl IntoResponse {
+    let mut headers = HeaderMap::new();
+
+    if TokenClaims::from_token(&params.token, &state.config.jwt_secret).is_ok() {
+        let cookie = format!("agamaToken={}; HttpOnly", params.token);
+        headers.insert(
+            SET_COOKIE,
+            cookie.parse().expect("could not build a valid cookie"),
+        );
+    }
+
+    headers.insert(LOCATION, "/".parse().unwrap());
+    (StatusCode::PERMANENT_REDIRECT, headers)
```

Comment:
Permanent redirect means that the browser should remember that and the future requests should directly go to the redirected location skipping the original request.

I'd rather use `StatusCode::TEMPORARY_REDIRECT` here because the redirection has a side effect of setting the cookie. If the browser is restarted or reloaded for whatever reason later with the `/login` path it could cause some nasty surprise...

Reply:
Good point!